src/java.net.http/share/classes/jdk/internal/net/http/websocket/Frame.java

```diff
@@ -132,18 +133,20 @@ Masker mask(int value) {
          * The source and the destination buffers may be the same instance. If
          * the mask hasn't been previously set it is assumed to be 0.
          */
-        Masker transferMasking(ByteBuffer src, ByteBuffer dst) {
-            begin(src, dst);
-            loop(src, dst);
-            end(src, dst);
-            return this;
+        public void applyMask(ByteBuffer src, ByteBuffer dst) {
+            if (src.order() == dst.order()) {
+                initVectorMask(src, dst);
+                applyVectorMask(src, dst);
+            }
+            applyPlainMask(src, dst);
         }
 
-        /*
-         * Applies up to 3 remaining from the previous pass bytes of the mask.
+        /**
+         * Positions the {@link #offset} at 0, which is needed for vectorized masking, by masking necessary amount of bytes.
          */
-        private void begin(ByteBuffer src, ByteBuffer dst) {
-            if (offset == 0) { // No partially applied mask from the previous invocation
+        private void initVectorMask(ByteBuffer src, ByteBuffer dst) {
```

Comment:
This method and `applyPlainMask` uses big-endian `maskBytes` which can be wrong if `dst` is not big endian. Should we just assert `dst` is big endian everywhere, as it seems to be the case?

Reply:
Opened intellij and verified the `dst` is always big endian coming from `ByteBuffer.allocate` in both client and server contexts.

Reply:
@liach, `initVectorMask()` operates byte-by-byte, that is, nothing vectorized, and, hence, no endianness concerns there. `applyVectorMask()` chooses the mask with correct endianness based on the input:

```
assert src.order() == dst.order() : "vectorized masking is only allowed on matching byte orders";
long maskLong = ByteOrder.LITTLE_ENDIAN == src.order() ? maskLongLe : maskLongBe;
```

AFAICT, both methods are ready to perform vectorization independent of the input endianness – granted `src` and `dst` endiannesses match. Am I missing something?

Reply:
For this line below:
```java
dst.put(j, (byte) (src.get(i) ^ maskBytes[offset]));
```
Because `maskBytes` is big-endian, if `dst` is little-endian (which is not the case at all right now because trusted callers are all using BE dst ByteBuffer), we should use `maskBytes[4 - offset]`, right?

Reply:
(Resolving this conversation, since it continues in a subsequent thread.)